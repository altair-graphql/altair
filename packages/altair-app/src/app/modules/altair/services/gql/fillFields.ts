import { visit, print, TypeInfo, parse, GraphQLSchema, Kind } from 'graphql';
import { debug } from '../../utils/logger';
import getTypeInfo from 'codemirror-graphql/utils/getTypeInfo';
import { Token } from 'codemirror';
import { ContextToken } from 'graphql-language-service';
import { buildSelectionSet } from './generateQuery';
import { Position } from '../../utils/editor/helpers';

export const parseQuery = (query: string) => {
  if (!query) {
    return null;
  }
  try {
    return parse(query);
  } catch (err) {
    debug.error('Something wrong with your query', err);

    return null;
  }
};

export const getIndentation = (str: string, index: number) => {
  let indentStart = index;
  let indentEnd = index;
  while (indentStart) {
    const c = str.charCodeAt(indentStart - 1);
    // line break
    if (c === 10 || c === 13 || c === 0x2028 || c === 0x2029) {
      break;
    }
    indentStart--;
    // not white space
    if (c !== 9 && c !== 11 && c !== 12 && c !== 32 && c !== 160) {
      indentEnd = indentStart;
    }
  }
  return str.substring(indentStart, indentEnd);
};

export const withInsertions = (
  initialQuery: string,
  insertions: { index: number; string: string }[]
) => {
  if (insertions.length === 0) {
    return initialQuery;
  }
  let edited = '';
  let prevIndex = 0;
  insertions.forEach(({ index, string }) => {
    edited += initialQuery.slice(prevIndex, index) + string;
    prevIndex = index;
  });
  edited += initialQuery.slice(prevIndex);
  return edited;
};

export interface FillAllFieldsOptions {
  maxDepth?: number;
}

// Improved version based on:
// https://github.com/graphql/graphiql/blob/272e2371fc7715217739efd7817ce6343cb4fbec/src/utility/fillLeafs.js
export const fillAllFields = (
  schema: GraphQLSchema,
  query: string,
  cursor: Position,
  token: ContextToken,
  { maxDepth = 1 }: FillAllFieldsOptions = {}
) => {
  const insertions: { index: number; string: string }[] = [];
  if (!schema) {
    return { insertions, result: query };
  }

  let tokenState = token.state as any;
  if (tokenState.kind === Kind.SELECTION_SET) {
    tokenState.wasSelectionSet = true;
    tokenState = { ...tokenState, ...tokenState.prevState };
  }
  const fieldType = getTypeInfo(schema, token.state).type;
  // Strip out empty selection sets since those throw errors while parsing query
  query = query.replace(/{\s*}/g, '');
  const ast = parseQuery(query);

  if (!fieldType || !ast) {
    return { insertions, result: query };
  }

  const typeInfo = new TypeInfo(schema);
  const edited = visit(ast, {
    leave(node) {
      // typeInfo.leave(node);
      debug.log(node);
    },
    enter(node) {
      typeInfo.enter(node);
      if (node.kind === Kind.FIELD) {
        // const fieldType = typeInfo.getType();
        if (
          1 &&
          node.name.value === tokenState.name &&
          node.loc &&
          // With tokenState of SelectionSet (i.e. between braces { | }),
          // we wouldn't have the right position since we strip out the empty braces before
          // parsing the query to make it valid
          // AST line number is 1-indexed while codemirror cursor line number is 0-indexed.
          (tokenState.wasSelectionSet ||
            node.loc.startToken.line - 1 === cursor.line)
        ) {
          debug.log(
            node,
            typeInfo,
            typeInfo.getType(),
            cursor,
            token,
            maxDepth
          );
          const { selectionSet } = buildSelectionSet(fieldType, { maxDepth });
          const indent = getIndentation(query, node.loc.start);
          if (selectionSet) {
            insertions.push({
              index: node.loc.end,
              string: ' ' + print(selectionSet).replace(/\n/g, '\n' + indent),
            });
          }
          return {
            ...node,
            selectionSet,
          };
        }
      }
    },
  });

  return {
    insertions,
    result: withInsertions(query, insertions),
  };
};
