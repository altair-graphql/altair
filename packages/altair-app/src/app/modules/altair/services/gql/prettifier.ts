export const prettify = async (query: string, tabWidth = 2) => {
  const prettier = await import('prettier/standalone');
  const prettierGraphql = await import('prettier/plugins/graphql');
  // return print(parse(query));
  return prettier.format(query, {
    parser: 'graphql',
    plugins: [prettierGraphql.default],
    tabWidth,
  });
};
