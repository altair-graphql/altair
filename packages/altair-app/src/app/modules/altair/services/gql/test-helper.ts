import { readFileSync } from 'fs';
import { buildSchema } from 'graphql';
import { resolve } from 'path';

export const getTestSchema = () => {
  const sdl = readFileSync(
    resolve(__dirname, '../../../../../../fixtures/test.sdl'),
    'utf8'
  );
  return buildSchema(sdl);
};
