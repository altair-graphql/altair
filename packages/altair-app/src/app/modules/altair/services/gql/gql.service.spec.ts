import { TestBed, inject } from '@angular/core/testing';
import { expect, jest } from '@jest/globals';

import { HttpClient, HttpResponse, HttpErrorResponse } from '@angular/common/http';

import { GqlService } from './gql.service';
import { NotifyService } from '../notify/notify.service';
import { Store } from '@ngrx/store';
import { EMPTY, of, throwError } from 'rxjs';
import { take } from 'rxjs/operators';
import { IntrospectionQuery, buildClientSchema } from 'graphql';

import validIntrospectionData from './__mock__/valid-introspection-data';
import { anyFn, mock } from '../../../../../testing';
import { RootState } from 'altair-graphql-core/build/types/state/state.interfaces';
import { Position } from '../../utils/editor/helpers';
import { ElectronAppService } from '../electron-app/electron-app.service';
import { MockProvider } from 'ng-mocks';
import { GraphQLRequestHandler } from 'altair-graphql-core/build/request/types';
import { PerWindowState } from 'altair-graphql-core/build/types/state/per-window.interfaces';

let mockHttpClient: HttpClient;
let mockNotifyService: NotifyService;
let mockElectronAppService: ElectronAppService;
let mockStore: Store<RootState>;
let mockRequestHandler: GraphQLRequestHandler;

describe('GqlService', () => {
  beforeEach(() => {
    mockHttpClient = {} as HttpClient;
    mockHttpClient.request = jest.fn(() => {
      return EMPTY;
    });
    mockStore = {
      subscribe: anyFn(),
    } as Store<RootState>;
    mockNotifyService = {
      error: anyFn(),
      info: anyFn(),
    } as NotifyService;
    mockElectronAppService = mock();
    mockRequestHandler = mock({
      handle(request) {
        return EMPTY;
      },
      generateCurl(request) {
        throw new Error('Method not implemented.');
      },
    });
    TestBed.configureTestingModule({
      providers: [
        GqlService,
        {
          provide: HttpClient,
          useFactory: () => mockHttpClient,
        },
        {
          provide: NotifyService,
          useFactory: () => mockNotifyService,
        },
        MockProvider(ElectronAppService),
        {
          provide: Store,
          useFactory: () => mockStore,
        },
      ],
      teardown: { destroyAfterEach: false },
    });
  });

  it('should create successfully', inject([GqlService], (service: GqlService) => {
    expect(service).toBeTruthy();
  }));

  describe('.getSelectedOperationData()', () => {
    it('should return selectedOperation as null for single queries, without requesting user selection', inject(
      [GqlService],
      (service: GqlService) => {
        expect(
          service.getSelectedOperationData({
            query: `query abc { hi }`,
          })
        ).toEqual({
          requestSelectedOperationFromUser: false,
          operations: [expect.anything()],
          selectedOperation: null,
        });
      }
    ));

    it('should return selectedOperation as single query name for single queries, if selectIfOneOperation is set', inject(
      [GqlService],
      (service: GqlService) => {
        expect(
          service.getSelectedOperationData({
            query: `query abc { hi }`,
            selectIfOneOperation: true,
          })
        ).toEqual({
          requestSelectedOperationFromUser: false,
          operations: [expect.anything()],
          selectedOperation: 'abc',
        });
      }
    ));

    it('should return selectedOperation as null for multiple queries, but with requesting user selection', inject(
      [GqlService],
      (service: GqlService) => {
        expect(
          service.getSelectedOperationData({
            query: `query abc { hi } query bcd{ bye }`,
          })
        ).toEqual({
          requestSelectedOperationFromUser: true,
          operations: [expect.anything(), expect.anything()],
          selectedOperation: null,
        });
      }
    ));

    it('should return provided selectedOperation if it matches any of the available queries', inject(
      [GqlService],
      (service: GqlService) => {
        expect(
          service.getSelectedOperationData({
            query: `query abc { hi } query bcd{ bye }`,
            selectedOperation: 'bcd',
          })
        ).toEqual({
          requestSelectedOperationFromUser: false,
          operations: [expect.anything(), expect.anything()],
          selectedOperation: 'bcd',
        });
      }
    ));
  });

  describe('.sendRequest()', () => {
    it('should call HttpClient with expected parameters', inject(
      [GqlService],
      (service: GqlService) => {
        service.sendRequest({
          url: 'http://test.com',
          method: 'post',
          query: '{}',
          extensions: '{}',
          additionalParams: '',
        });
        expect(mockHttpClient.request).toHaveBeenCalled();
        const httpClientArgs = (mockHttpClient.request as any).mock.calls[0];
        expect(httpClientArgs[0]).toBe('post');
        expect(httpClientArgs[1]).toBe('http://test.com');
        const httpConfigArg = httpClientArgs[2];
        expect(JSON.parse(httpConfigArg.body)).toEqual({
          query: '{}',
          variables: {},
          extensions: {},
          operationName: null,
        });
        expect(httpConfigArg.headers.get('Content-Type')).toBe('application/json');
      }
    ));

    it('should call HttpClient with correct content type when file is included', inject(
      [GqlService],
      (service: GqlService) => {
        service.sendRequest({
          url: 'http://test.com',
          method: 'post',
          query: '{}',
          files: [
            {
              name: 'file1',
              data: new File([], 'file1'),
            },
            {
              name: 'file2',
              data: new File([], 'file2'),
            },
          ],
          additionalParams: '',
        });
        expect(mockHttpClient.request).toHaveBeenCalled();
        const httpClientArgs = (mockHttpClient.request as any).mock.calls[0];
        expect(httpClientArgs[0]).toBe('post');
        expect(httpClientArgs[1]).toBe('http://test.com');
        const httpConfigArg = httpClientArgs[2];
        expect(httpConfigArg.body).toEqual(expect.any(FormData));
        expect(httpConfigArg.headers.get('Content-Type')).toBeFalsy();
        expect(JSON.parse(httpConfigArg.body.get('operations'))).toEqual({
          query: '{}',
          variables: {
            file1: null,
            file2: null,
          },
          operationName: null,
        });
      }
    ));

    it('should call HttpClient with default json content type if file passed to it is not valid', inject(
      [GqlService],
      (service: GqlService) => {
        service.sendRequest({
          url: 'http://test.com',
          method: 'post',
          query: '{}',
          files: [
            {
              name: 'file1',
            },
          ],
          additionalParams: '',
        });
        expect(mockHttpClient.request).toHaveBeenCalled();
        const httpClientArgs = (mockHttpClient.request as any).mock.calls[0];
        expect(httpClientArgs[0]).toBe('post');
        expect(httpClientArgs[1]).toBe('http://test.com');
        const httpConfigArg = httpClientArgs[2];
        expect(httpConfigArg.headers.get('Content-Type')).toBe('application/json');
        expect(JSON.parse(httpConfigArg.body)).toEqual({
          query: '{}',
          variables: {},
          operationName: null,
        });
      }
    ));
  });

  describe('.getIntrospectionRequest()', () => {
    it('should return introspection data', inject(
      [GqlService],
      async (service: GqlService) => {
        let httpClientCallCount = 0;
        mockRequestHandler.handle = (request) => {
          httpClientCallCount++;
          expect(request.selectedOperation).toEqual('IntrospectionQuery');
          switch (httpClientCallCount) {
            case 1: {
              return of({
                ok: true,
                data: JSON.stringify({
                  data: 'introspection data',
                }),
                headers: new Headers(),
                status: 200,
                statusText: 'OK',
                url: 'http://test.com',
                requestStartTimestamp: 1,
                requestEndTimestamp: 2,
                resopnseTimeMs: 1.5,
              });
            }
            default:
              return of({
                ok: true,
                data: JSON.stringify({
                  data: '',
                }),
                headers: new Headers(),
                status: 200,
                statusText: 'OK',
                url: 'http://test.com',
                requestStartTimestamp: 1,
                requestEndTimestamp: 2,
                resopnseTimeMs: 1.5,
              });
          }
        };
        const res = await service
          .getIntrospectionRequest({
            url: 'http://test.com',
            method: 'GET',
            additionalParams: '',
            handler: mockRequestHandler,
          })
          .pipe(take(1))
          .toPromise();

        expect(res?.body).toEqual(
          JSON.stringify({
            data: 'introspection data',
          })
        );
      }
    ));

    it('should fetch introspection with old introspection query if first call throws error', inject(
      [GqlService],
      async (service: GqlService) => {
        let httpClientCallCount = 0;
        mockRequestHandler.handle = jest.fn(() => {
          httpClientCallCount++;
          switch (httpClientCallCount) {
            case 1: {
              return throwError(() => new Error('Some network error'));
            }
            case 2: {
              return of({
                ok: true,
                data: JSON.stringify({
                  data: 'second introspection data',
                }),
                headers: new Headers(),
                status: 200,
                statusText: 'OK',
                url: 'http://test.com',
                requestStartTimestamp: 1,
                requestEndTimestamp: 2,
                resopnseTimeMs: 1.5,
              });
            }
            default:
              return of({
                ok: true,
                data: JSON.stringify({
                  data: '',
                }),
                headers: new Headers(),
                status: 200,
                statusText: 'OK',
                url: 'http://test.com',
                requestStartTimestamp: 1,
                requestEndTimestamp: 2,
                resopnseTimeMs: 1.5,
              });
          }
        });
        const res = await service
          .getIntrospectionRequest({
            url: 'http://test.com',
            method: 'GET',
            additionalParams: '',
            handler: mockRequestHandler,
          })
          .pipe(take(1))
          .toPromise();

        expect(res?.body).toEqual(
          JSON.stringify({
            data: 'second introspection data',
          })
        );
      }
    ));

    it('should throw error if second attempt also fails', inject(
      [GqlService],
      async (service: GqlService) => {
        let httpClientCallCount = 0;
        mockHttpClient.request = (...args: any) => {
          httpClientCallCount++;
          switch (httpClientCallCount) {
            case 1: {
              const resp = new HttpErrorResponse({
                error: 'Some network error',
              });
              return of(resp) as any;
            }
            case 2: {
              const resp = new HttpErrorResponse({
                error: 'Second network error',
              });
              return of(resp) as any;
            }
            default:
              return of(
                new HttpResponse<any>({
                  body: {
                    data: '',
                  },
                })
              ) as any;
          }
        };

        try {
          const res = await service
            .getIntrospectionRequest({
              url: 'http://test.com',
              method: 'GET',
            })
            .pipe(take(1))
            .toPromise();
          // Should not be called
          expect(true).toBe(false);
        } catch (err) {
          expect(true).toBe(true);
        }
      }
    ));
  });

  describe('.setHeaders()', () => {
    it('should set headers with default headers', inject(
      [GqlService],
      async (service: GqlService) => {
        service.setHeaders();

        const headers = service.headers;

        expect(headers.get('Content-Type')).toBe('application/json');
        expect(headers.get('Accept')).toBe('application/json');
      }
    ));

    it('should set only enabled headers with default headers', inject(
      [GqlService],
      async (service: GqlService) => {
        service.setHeaders([
          { key: 'x-header-1', value: 'Header 1', enabled: true },
          { key: 'x-header-2', value: 'Header 2', enabled: false },
          { key: 'x-header-3', value: 'Header 3', enabled: true },
        ]);

        const headers = service.headers;

        expect(headers.get('Content-Type')).toBe('application/json');
        expect(headers.get('Accept')).toBe('application/json');
        expect(headers.get('x-header-1')).toBe('Header 1');
        expect(headers.get('x-header-2')).toBe(null);
        expect(headers.get('x-header-3')).toBe('Header 3');
      }
    ));

    it('should NOT set forbidden headers', inject(
      [GqlService],
      async (service: GqlService) => {
        service.setHeaders([
          { key: 'Origin', value: 'http://mysite.com', enabled: true },
        ]);

        const headers = service.headers;

        expect(headers.get('Origin')).toBe(null);
      }
    ));

    it('should NOT merge default headers is skipDefault is true', inject(
      [GqlService],
      async (service: GqlService) => {
        service.setHeaders(
          [
            { key: 'x-header-1', value: 'Header 1', enabled: true },
            { key: 'x-header-3', value: 'Header 3', enabled: true },
          ],
          { skipDefaults: true }
        );

        const headers = service.headers;

        expect(headers.get('Content-Type')).toBe(null);
        expect(headers.get('Accept')).toBe(null);
        expect(headers.get('x-header-1')).toBe('Header 1');
        expect(headers.get('x-header-3')).toBe('Header 3');
      }
    ));
  });

  describe('.getIntrospectionSchema()', () => {
    it('should return null if no introspection data is provided', inject(
      [GqlService],
      async (service: GqlService) => {
        const schema = service.getIntrospectionSchema();

        expect(schema).toBe(null);
      }
    ));

    it('should return schema for introspection data', inject(
      [GqlService],
      async (service: GqlService) => {
        const schema = service.getIntrospectionSchema(validIntrospectionData as any);

        expect(schema).toMatchSnapshot();
      }
    ));
    it('should return null if it cannot parse introspection data', inject(
      [GqlService],
      async (service: GqlService) => {
        const schema = service.getIntrospectionSchema({
          __schema: {
            queryType: {
              name: 'Query',
            },
          },
        } as IntrospectionQuery);

        expect(schema).toBe(null);
      }
    ));
  });

  describe('.hasInvalidFileVariable()', () => {
    it('should return false if no file variables are passed', inject(
      [GqlService],
      (service: GqlService) => {
        expect(service.hasInvalidFileVariable([])).toBe(false);
      }
    ));

    it('should return true if any file variables has no file data', inject(
      [GqlService],
      (service: GqlService) => {
        expect(service.hasInvalidFileVariable([{ name: 'file1' }])).toBe(true);
      }
    ));

    it('should return true if file data of file variables is not a File', inject(
      [GqlService],
      (service: GqlService) => {
        expect(
          service.hasInvalidFileVariable([{ name: 'file1', data: {} as any }])
        ).toBe(true);
      }
    ));

    it('should return true if any file variables has no file name', inject(
      [GqlService],
      (service: GqlService) => {
        expect(
          service.hasInvalidFileVariable([{ name: '', data: new File([], 'file') }])
        ).toBe(true);
      }
    ));

    it('should return false if all file variables are valid', inject(
      [GqlService],
      (service: GqlService) => {
        expect(
          service.hasInvalidFileVariable([
            { name: 'file1', data: new File([], 'file') },
          ])
        ).toBe(false);
      }
    ));
  });

  describe('.fillAllFields()', () => {
    it('should return query with fields filled', inject(
      [GqlService],
      (service: GqlService) => {
        const schema = buildClientSchema(validIntrospectionData as any);
        const query = `
        query {
          GOTHouses(name: "first"){
          }
        }
      `;
        const token: any = {
          start: 10,
          end: 11,
          string: '}',
          state: {
            level: 0,
            step: 1,
            name: null,
            kind: 'SelectionSet',
            type: null,
            rule: [
              {
                style: 'punctuation',
              },
              {
                ofRule: 'Selection',
                isList: true,
              },
              {
                style: 'punctuation',
              },
            ],
            needsSeperator: false,
            prevState: {
              level: 0,
              step: 3,
              name: 'GOTHouses',
              kind: 'Field',
              type: null,
              rule: [
                {
                  style: 'property',
                },
                {
                  ofRule: 'Arguments',
                },
                {
                  ofRule: 'Directive',
                  isList: true,
                },
                {
                  ofRule: 'SelectionSet',
                },
              ],
              needsSeperator: false,
              prevState: {
                level: 0,
                step: 0,
                name: null,
                kind: 'Selection',
                type: null,
                needsSeperator: false,
                prevState: {
                  level: 0,
                  step: 1,
                  name: null,
                  kind: 'SelectionSet',
                  type: null,
                  rule: [
                    {
                      style: 'punctuation',
                    },
                    {
                      ofRule: 'Selection',
                      isList: true,
                    },
                    {
                      style: 'punctuation',
                    },
                  ],
                  needsSeperator: false,
                  prevState: {
                    level: 0,
                    step: 4,
                    name: null,
                    kind: 'Query',
                    type: null,
                    rule: [
                      {
                        style: 'keyword',
                      },
                      {
                        ofRule: {
                          style: 'def',
                        },
                      },
                      {
                        ofRule: 'VariableDefinitions',
                      },
                      {
                        ofRule: 'Directive',
                        isList: true,
                      },
                      'SelectionSet',
                    ],
                    needsSeperator: false,
                    prevState: {
                      level: 0,
                      step: 0,
                      name: null,
                      kind: 'Definition',
                      type: null,
                      needsSeperator: false,
                      prevState: {
                        level: 0,
                        step: 0,
                        name: null,
                        kind: 'Document',
                        type: null,
                        rule: [
                          {
                            ofRule: 'Definition',
                            isList: true,
                          },
                        ],
                        needsSeperator: false,
                        prevState: {
                          level: 0,
                          step: 0,
                          name: null,
                          kind: null,
                          type: null,
                          rule: null,
                          needsSeperator: false,
                          prevState: null,
                        },
                        indentLevel: 0,
                      },
                      indentLevel: 0,
                    },
                    indentLevel: 0,
                    needsAdvance: false,
                    levels: [1],
                  },
                  indentLevel: 1,
                  needsAdvance: false,
                  levels: [1],
                },
                indentLevel: 1,
                needsAdvance: false,
                levels: [1],
              },
              indentLevel: 1,
              needsAdvance: false,
              levels: [1, 2],
            },
            indentLevel: 1,
            needsAdvance: false,
            levels: [1, 2],
            wasSelectionSet: true,
          },
          type: 'punctuation',
        };
        const res = service.fillAllFields(schema, query, new Position(3, 0), token, {
          maxDepth: 1,
        });

        expect(res).toMatchSnapshot();
      }
    ));
  });

  describe('.parseQuery()', () => {
    it('should return empty object if empty string', inject(
      [GqlService],
      (service: GqlService) => {
        expect(service.parseQueryOrEmptyDocument('')).toEqual({
          definitions: [],
          kind: 'Document',
        });
      }
    ));

    it('should return GraphQL document', inject(
      [GqlService],
      (service: GqlService) => {
        expect(
          service.parseQuery(`
        query {
          hello
        }
      `)
        ).toMatchSnapshot();
      }
    ));

    it('should return empty object if query is invalid', inject(
      [GqlService],
      (service: GqlService) => {
        expect(
          service.parseQueryOrEmptyDocument(`
        query {{
          hello
        }
      `)
        ).toEqual({
          definitions: [],
          kind: 'Document',
        });
      }
    ));
  });

  describe('.isSchema()', () => {
    it('should return true if argument is a GraphQL schema', inject(
      [GqlService],
      (service: GqlService) => {
        expect(
          service.isSchema(buildClientSchema(validIntrospectionData as any))
        ).toBe(true);
      }
    ));

    it('should return false if argument is not a GraphQL schema', inject(
      [GqlService],
      (service: GqlService) => {
        expect(service.isSchema({})).toBe(false);
      }
    ));
  });

  describe('.isSubscriptionQuery()', () => {
    it('should return true if query is a subscription query', inject(
      [GqlService],
      (service: GqlService) => {
        expect(
          service.isSubscriptionQuery('subscription { get { id message } }', {
            query: {
              queryEditorState: {},
            },
          } as unknown as PerWindowState)
        ).toBe(true);
      }
    ));

    it('should return false if query is not a subscription query', inject(
      [GqlService],
      (service: GqlService) => {
        expect(
          service.isSubscriptionQuery('query { get { id message } }', {
            query: {},
          } as unknown as PerWindowState)
        ).toBe(false);
      }
    ));
  });

  describe('.getOperationNameAtIndex()', () => {
    it('should return name of graphql operation at index position', inject(
      [GqlService],
      (service: GqlService) => {
        expect(
          service.getOperationNameAtIndex(`query first { get { id message } }`, 2)
        ).toBe('first');
      }
    ));

    it('should return name of graphql operation at index position for multiple queries', inject(
      [GqlService],
      (service: GqlService) => {
        expect(
          service.getOperationNameAtIndex(
            `
        query first { get { id message } }
        query second { use { id } }
      `,
            56
          )
        ).toBe('second');
      }
    ));
  });

  describe('.nameQuery()', () => {
    it('should return edited query with generated name', inject(
      [GqlService],
      (service: GqlService) => {
        expect(service.nameQuery(`query { get { id message } }`)).toMatch(
          /query [A-Za-z0-9_]+ {/
        );
      }
    ));
  });

  describe('.refactorQuery()', () => {
    it('should return edited query with generated name', inject(
      [GqlService],
      (service: GqlService) => {
        const _originalRandom = Math.random;
        let cnt = 1;
        Math.random = jest.fn(() => cnt++);
        const res = service.refactorQuery(
          `
        query{
          GOTHouses(name: "first"){
            id
            url
            name
            region
            titles
            seats
            words
          }
        }
        query{
          GOTHouses(name: "second"){
            id
            url
            name
            region
            titles
            seats
            words
          }
        }
      `,
          buildClientSchema(validIntrospectionData as any)
        );

        expect(res?.query).toMatchSnapshot();
        expect(res?.variables).toMatchSnapshot();

        Math.random = _originalRandom;
      }
    ));
  });

  describe('.getSDLSync()', () => {
    it('should return empty string for non-schema', inject(
      [GqlService],
      (service: GqlService) => {
        const sdl = service.getSDLSync({} as any);

        expect(sdl).toBe('');
      }
    ));
    it('should return SDL for schema', inject(
      [GqlService],
      (service: GqlService) => {
        const sdl = service.getSDLSync(
          buildClientSchema(validIntrospectionData as any)
        );

        expect(sdl).toBeTruthy();
      }
    ));
  });

  describe('.createStreamClient()', () => {
    it('should return event source', inject([GqlService], (service: GqlService) => {
      const client = service.createStreamClient('http://example.com/stream');

      expect(client.url).toBe('http://example.com/stream');
      expect(client).toEqual(expect.any(EventSource));
    }));
  });

  describe('.closeStreamClient()', () => {
    it('should close client', inject([GqlService], (service: GqlService) => {
      const client = new EventSource('http://example.com/stream');
      const closeSpy = jest.spyOn(client, 'close');
      service.closeStreamClient(client);

      expect(client.close).toHaveBeenCalled();
    }));
  });

  describe('.normalizeFiles()', () => {
    it('should return a list of graphql multipart file upload compliant file variables', inject(
      [GqlService],
      (service: GqlService) => {
        const files = [
          {
            name: 'first',
            data: new File([''], 'first.file'),
          },
          {
            name: 'second',
            isMultiple: true,
            data: [
              new File([''], 'second1.file'),
              new File([''], 'second2.file'),
              {} as File,
            ],
          },
          {
            name: 'third',
            data: {} as File,
          },
          {
            name: 'fourth.0',
            data: new File([''], 'fourth.file'),
          },
          {
            name: '',
            data: new File([''], 'unknown.file'),
          },
        ];
        const result = service.normalizeFiles(files);

        // expect(result).toBe('');
        expect(result.resolvedFiles.length).toBe(4);
        expect(result.resolvedFiles).toEqual(
          expect.arrayContaining([
            {
              name: 'first',
              data: expect.any(File),
            },
            {
              name: 'second.0',
              data: expect.any(File),
            },
            {
              name: 'second.1',
              data: expect.any(File),
            },
            {
              name: 'fourth.0',
              data: expect.any(File),
            },
          ])
        );
        expect(result.erroneousFiles).toEqual(
          expect.arrayContaining([
            {
              name: 'second.2',
              data: expect.any(Object),
            },
            {
              name: 'third',
              data: expect.any(Object),
            },
            {
              name: '',
              data: expect.any(File),
            },
          ])
        );
      }
    ));
    it('should return only the first file data if not multiple and has > 1 data', inject(
      [GqlService],
      (service: GqlService) => {
        const files = [
          {
            name: 'first',
            data: new File([''], 'first.file'),
          },
          {
            name: 'second',
            data: [new File([''], 'second1.file')],
          },
          {
            name: 'third',
            data: [new File([''], 'third1.file'), new File([''], 'third2.file')],
          },
        ];
        const result = service.normalizeFiles(files);

        // expect(result).toBe('');
        expect(result.resolvedFiles.length).toBe(3);
        expect(result.resolvedFiles).toEqual(
          expect.arrayContaining([
            {
              name: 'first',
              data: expect.any(File),
            },
            {
              name: 'second',
              data: expect.any(File),
            },
            {
              name: 'third',
              data: expect.any(File),
            },
          ])
        );
        expect(result.erroneousFiles.length).toEqual(0);
      }
    ));
  });
});
