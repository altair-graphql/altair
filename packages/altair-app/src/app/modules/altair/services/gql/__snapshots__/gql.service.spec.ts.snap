// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GqlService .fillAllFields() should return query with fields filled 1`] = `
{
  "insertions": [
    {
      "index": 51,
      "string": " {
            id
            url
            name
            region
            titles
            seats
            words
            overlord
            swornMembers
          }",
    },
  ],
  "result": "
        query {
          GOTHouses(name: "first") {
            id
            url
            name
            region
            titles
            seats
            words
            overlord
            swornMembers
          }
        }
      ",
}
`;

exports[`GqlService .getIntrospectionSchema() should return schema for introspection data 1`] = `
GraphQLSchema {
  "__validationErrors": undefined,
  "_directives": [
    "@cacheControl",
    "@skip",
    "@include",
    "@deprecated",
  ],
  "_implementationsMap": {},
  "_mutationType": "Mutation",
  "_queryType": "Query",
  "_subTypeMap": {},
  "_subscriptionType": "Subscription",
  "_typeMap": {
    "Boolean": "Boolean",
    "CacheControlScope": "CacheControlScope",
    "File": "File",
    "GOTBook": "GOTBook",
    "GOTCharacter": "GOTCharacter",
    "GOTHouse": "GOTHouse",
    "HNUser": "HNUser",
    "Int": "Int",
    "Message": "Message",
    "Mutation": "Mutation",
    "Query": "Query",
    "String": "String",
    "Subscription": "Subscription",
    "Upload": "Upload",
    "__Directive": "__Directive",
    "__DirectiveLocation": "__DirectiveLocation",
    "__EnumValue": "__EnumValue",
    "__Field": "__Field",
    "__InputValue": "__InputValue",
    "__Schema": "__Schema",
    "__Type": "__Type",
    "__TypeKind": "__TypeKind",
  },
  "astNode": undefined,
  "description": undefined,
  "extensionASTNodes": undefined,
  "extensions": undefined,
}
`;

exports[`GqlService .parseQuery() should return GraphQL document 1`] = `
{
  "definitions": [
    {
      "directives": [],
      "kind": "OperationDefinition",
      "loc": {
        "end": 42,
        "start": 9,
      },
      "name": undefined,
      "operation": "query",
      "selectionSet": {
        "kind": "SelectionSet",
        "loc": {
          "end": 42,
          "start": 15,
        },
        "selections": [
          {
            "alias": undefined,
            "arguments": [],
            "directives": [],
            "kind": "Field",
            "loc": {
              "end": 32,
              "start": 27,
            },
            "name": {
              "kind": "Name",
              "loc": {
                "end": 32,
                "start": 27,
              },
              "value": "hello",
            },
            "selectionSet": undefined,
          },
        ],
      },
      "variableDefinitions": [],
    },
  ],
  "kind": "Document",
  "loc": {
    "end": 49,
    "start": 0,
  },
}
`;

exports[`GqlService .refactorQuery() should return edited query with generated name 1`] = `
"query refactored1000($name: String) {
  GOTHouses(name: $name) {
    ...GOTHouseFields
  }
}

query refactored3000($name2000: String) {
  GOTHouses(name: $name2000) {
    ...GOTHouseFields
  }
}

fragment GOTHouseFields on GOTHouse {
  id
  url
  name
  region
  titles
  seats
  words
}
"
`;

exports[`GqlService .refactorQuery() should return edited query with generated name 2`] = `
{
  "name": "first",
  "name2000": "second",
}
`;
